network/peer-monitoring-service/server/src/lib.rs

```diff
@@ -180,61 +173,95 @@ impl Handler {
+        // Create and send the response
+        let network_information_response = NetworkInformationResponse {
+            connected_peers_and_metadata,
+            depth_from_validators,
+        };
+        Ok(PeerMonitoringServiceResponse::NetworkInformation(
+            network_information_response,
         ))
     }
 
     fn get_server_protocol_version(&self) -> Result<PeerMonitoringServiceResponse, Error> {
+        let server_protocol_version_response = ServerProtocolVersionResponse {
+            version: PEER_MONITORING_SERVER_VERSION,
+        };
         Ok(PeerMonitoringServiceResponse::ServerProtocolVersion(
-            ServerProtocolVersionResponse {
-                version: PEER_MONITORING_SERVER_VERSION,
-            },
+            server_protocol_version_response,
         ))
     }
 
-    fn get_validators_and_vfns(&self) -> Result<PeerMonitoringServiceResponse, Error> {
-        Err(Error::InvalidRequest(
-            "get_validators_and_vfns() is currently unsupported!".into(),
+    fn handle_latency_ping(
+        &self,
+        latency_ping_request: &LatencyPingRequest,
+    ) -> Result<PeerMonitoringServiceResponse, Error> {
+        let latency_ping_response = LatencyPingResponse {
+            ping_counter: latency_ping_request.ping_counter,
+        };
+        Ok(PeerMonitoringServiceResponse::LatencyPing(
+            latency_ping_response,
         ))
     }
+}
 
-    fn handle_ping(&self) -> Result<PeerMonitoringServiceResponse, Error> {
-        Err(Error::InvalidRequest(
-            "handle_ping() is currently unsupported!".into(),
-        ))
+/// Returns the depth from the validators using the given base config
+/// and the peers and metadata information.
+fn get_depth_from_validators(
+    base_config: &BaseConfig,
+    peers_and_metadata: Arc<PeersAndMetadata>,
+) -> u64 {
+    match base_config.role {
+        RoleType::Validator => 0, // We're a validator!
+        RoleType::FullNode => {
+            match peers_and_metadata.get_connected_peers_and_metadata() {
+                Ok(peers_and_metadata) => {
+                    // Go through our peers, find the min, and return a depth relative to the min
+                    let mut min_peer_depth_from_validators = MAX_DEPTH_FROM_VALIDATORS;
+                    for peer_metadata in peers_and_metadata.values() {
+                        if let Some(depth_from_validators) = peer_metadata
+                            .get_peer_monitoring_metadata()
+                            .depth_from_validators
+                        {
+                            min_peer_depth_from_validators =
+                                min(min_peer_depth_from_validators, depth_from_validators);
+                        }
+                    }
+
+                    // We're one hop away from the peer
+                    min(
+                        MAX_DEPTH_FROM_VALIDATORS,
+                        min_peer_depth_from_validators + 1,
```

Comment:
I guess we can cache this until peer connect and disconnect?

Reply:
Umm, I'm not sure. I think it's possible that our distance changes even if we don't get any new peer connects/disconnects (e.g., imagine a peer that we're connected to gets a new connection. This could improve our distance once we've seen that they've got a new distance). Also, given our previous discussions, I'm gonna leave a call caching until we see them become a problem 😄